pkg/controllers/webhookcerts/cert_requester.go

```diff
@@ -0,0 +1,398 @@
+					klog.Warning("the previous CSR got denied or it otherwise failed: %v", cond)
+					if err := c.resubmitCSR(ctx, key); err != nil {
+						return nil, err
+					}
+					return pbool(true), nil
+				}
+			}
+		}
+
+		for _, cond := range csrObj.Status.Conditions {
+			if cond.Type == certapiv1.CertificateApproved && cond.Status == corev1.ConditionTrue {
+				csrCert := csrObj.Status.Certificate
+				if len(csrCert) == 0 {
+					return pbool(true), nil
+				}
+
+				certObj, err := pemDecodeCert(csrCert)
+				if err != nil {
+					return nil, fmt.Errorf("failed to parse the certiticate retrieved from CSR: %w", err)
+				}
+
+				if !key.PublicKey.Equal(certObj.PublicKey) {
+					syncCtx.Recorder().Eventf("CSRBadCertKey", "the public key of the CSR certificate did not match the current private key")
+					klog.Warning("the public key of the CSR certificate did not match the current private key")
+					if err := c.resubmitCSR(ctx, key); err != nil {
+						return nil, err
+					}
+					return pbool(true), nil
+				}
+
+				certSecret, err := c.secretLister.Secrets("openshift-oauth-apiserver").Get(certSecretName)
+				if err != nil {
+					return nil, err
+				}
+				certSecret.Data["tls.crt"] = csrCert
+				_, err = c.secretClient.Update(ctx, certSecret, metav1.UpdateOptions{})
+				if err != nil {
+					return nil, err
+				}
+				// we've got the cert, stop progressing
+				return pbool(false), nil
+			}
+		}
+
+		// no conditions with "True" status
+		return pbool(true), nil
+
+	}
+
+	if err = c.resubmitCSR(ctx, key); err != nil {
+		return nil, err
+	}
+
+	return pbool(true), nil
+}
+
+func (c *webhookAuthenticatorCertRequester) getKeyCertFromSecret(ctx context.Context) (key *rsa.PrivateKey, cert *x509.Certificate, keyRotated bool, err error) {
+	certSecret, err := c.secretLister.Secrets("openshift-oauth-apiserver").Get(certSecretName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, nil, false, err
+	}
+
+	var keyPEM []byte
+	if certSecret != nil {
+		keyPEM = certSecret.Data["tls.key"]
+	}
+	if len(keyPEM) > 0 {
+		key, err = pemDecodeRSAPrivate(keyPEM)
+		if err != nil {
+			klog.Error("failed to decode the private key used for the openshift-authenticator, new key will be generated: %v", err)
+			// delete the secret and regenerate it in the next step with a new key
+			err = c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{})
+			if err != nil {
+				return nil, nil, false, err
+			}
+		}
+	}
+	if key == nil {
+		key, err = rsa.GenerateKey(rand.Reader, 4096)
+		if err != nil {
+			return nil, nil, false, fmt.Errorf("failed to generate a private key: %w", err)
+		}
+		keyPEM, err := pemEncodeRSAPrivate(key)
+		if err != nil {
+			return nil, nil, false, err
+		}
+
+		if err := c.secretClient.Delete(ctx, certSecretName, metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
```

Comment:
is this `Delete` necessary? 

`key == nil` means that there was not secret.

Reply:
...or the secret was there but had `tls.key` empty